devdocs/troubleshooting-guide.md

```diff
@@ -0,0 +1,204 @@
+
+  **SOLUTION**: Eclipse may be using an incompatible version of the JRE Library (with respect to TEAMMATES) for the current JRE definition. System Library for JRE should be set to the workspace default, after appropriate JRE definition has been added (covered in existing point#2 of the troubleshooting section).
+  
+   
+   Right-click on project → Properties → Java Build Path → "Libraries" tab → Select JRE System Library from the list → Edit… → Select and mark radio button for "Workspace default JRE"
+   
+
+7. **ERROR**: Everything seems normal and the dev server starts normally. But you get a "Class not found" or "Class cannot initialize" error when you try to access any non-static pages of the app.
+
+  **SOLUTION**: Check if the build.properties and appengine-web.xml has been renamed properly, which you were supposed to during the set up.
+
+## Troubleshooting test failures
+
+### Optimizing IDE layout for testing
+
+  The default place for the TestNG tab is alongside the Console tab.
+
+  ![troubleshooting-test-1.png](images/troubleshooting-test-1.png)
+
+  Here's a better place for it. Just drag the tab and drop it alongside the Project Explorer tab.
+
+  ![troubleshooting-test-2.png](images/troubleshooting-test-2.png)
+
+### How automated browser tests work
+
+  Our automated Browser tests compare generated Web pages with expected pages (these are in the \src\test\resources\pages folder). Such a test case fails when the expected page does not match the actual page. To see the diff between actual and expected, double click the error message in the TestNG tab.
+  
+Here is an example (use the zoom feature of the Browser to see details of the screenshot).
+
+  ![troubleshooting-test-3.png](images/troubleshooting-test-3.png)
+
+  Note that you can ignore the following differences reported by Eclipse diff tool. Those are not the cause of the test failure.
+
+ Expected | Actual | Explanation 
+ --- | --- | --- 
+ `<expected></expected>` | `<actual></actual>` | These are inserted by the test code for easier reference, after the comparison has been done. 
+ {*} | Any text | The expression {*} is used to mark a piece of text we cannot predict in advance. e.g. current time. These differences are ignored by our comparison algorithm. 
+
+  Furthermore, you might see tags such as ${test.student1} and ${version} in some of the  expected files (e.g. StudentHomeHTML.html) if you open them in Eclipse. The test code will replace those tags with correct values before performing the comparison. For example, ${version} will be replaced by something like V4.89
+
+### Getting help to resolve test failures
+
+Remember to state which tests are failing and for each of those
+* The detailed error message (taken from the TestNG tab)
+
+* Any relevant screenshots (e.g. screen shot of the diff window showing the offending difference)
+
+### Other errors and solutions
+
+1. **ERROR**: Tests fail randomly during dev server testing.
```

Comment:
Add two more items here as number 1 and 2:

1. Error: Encountered "java.net.ConnectException: Connection refused" when running some tests.
    Solution: Ensure that your dev server is started prior to running those tests.

2. Error: Encountered "org.openqa.selenium.WebDriverException: Unable to bind to locking port 7054 within 45000 ms" when running tests with Browser.
    Solution: Ensure compatible version of Firefox is installed as specified under Prerequisites.